File: travel_expense_backend/src/modules/approvals/entities/approval-stage.entity.ts
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, UpdateDateColumn, ManyToOne, JoinColumn } from 'typeorm';
import { User } from '../../users/entities/user.entity';
import { TravelRequest } from '../../travel/entities/travel-request.entity';

export enum ApprovalStageStatus {
  PENDING = 'pending',
  APPROVED = 'approved',
  REJECTED = 'rejected',
  ESCALATED = 'escalated',
  SKIPPED = 'skipped',
}

@Entity('approval_stages')
export class ApprovalStage {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  requestId: string;

  @ManyToOne(() => TravelRequest, (tr) => tr.approvalStages, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'requestId' })
  travelRequest: TravelRequest;

  @Column()
  approverId: string;

  @ManyToOne(() => User, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'approverId' })
  approver: User;

  @Column()
  level: string; // E.g., 'L1_MANAGER', 'L2_DEPT_HEAD', 'L3_FINANCE'

  @Column({
    type: process.env.OFFLINE_MODE === 'true' ? 'simple-enum' : 'enum',
    enum: ApprovalStageStatus,
    default: ApprovalStageStatus.PENDING,
  })
  status: ApprovalStageStatus;

  @Column({ type: 'text', nullable: true })
  comments: string;

  @UpdateDateColumn()
  updatedAt: Date;

  @CreateDateColumn()
  createdAt: Date;
}

File: travel_expense_app/lib/features/approvals/domain/entities/approval_stage_entity.dart
import 'package:equatable/equatable.dart';
import '../../../travel_request/domain/entities/travel_request_entity.dart';
import '../../../auth/domain/entities/user_entity.dart';

class ApprovalStageEntity extends Equatable {
  final String id;
  final String requestId;
  final TravelRequestEntity? travelRequest;
  final String approverId;
  final UserEntity? approver;
  final String level;
  final String status;
  final String? comments;
  final DateTime createdAt;

  const ApprovalStageEntity({
    required this.id,
    required this.requestId,
    this.travelRequest,
    required this.approverId,
    this.approver,
    required this.level,
    required this.status,
    this.comments,
    required this.createdAt,
  });

  @override
  List<Object?> get props => [
        id,
        requestId,
        travelRequest,
        approverId,
        approver,
        level,
        status,
        comments,
        createdAt,
      ];
}
